internal/block/block_pool_test.go

```diff
@@ -0,0 +1,138 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package block
+
+import (
+	"fmt"
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+const invalidConfigError string = "invalid configuration provided for blockPool, blocksize: %d, maxBlocks: %d"
+
+type BlockPoolTest struct {
+	suite.Suite
+}
+
+func TestBlockPoolTestSuite(t *testing.T) {
+	suite.Run(t, new(BlockPoolTest))
+}
+
+func (t *BlockPoolTest) TestInitBlockPool() {
+	bp, err := InitBlockPool(1024, 10)
+
+	require.Nil(t.T(), err)
+	assert.NotNil(t.T(), bp)
+	assert.Equal(t.T(), int64(1024), bp.blockSize)
+	assert.Equal(t.T(), int32(10), bp.maxBlocks)
+	assert.Equal(t.T(), int32(0), bp.totalBlocks)
+}
+
+func (t *BlockPoolTest) TestInitBlockPoolForZeroBlockSize() {
+	_, err := InitBlockPool(0, 10)
+
+	require.NotNil(t.T(), err)
+	assert.Equal(t.T(), fmt.Errorf(invalidConfigError, 0, 10), err)
+}
+
+func (t *BlockPoolTest) TestInitBlockPoolForNegativeBlockSize() {
+	_, err := InitBlockPool(-1, 10)
+
+	require.NotNil(t.T(), err)
+	assert.Equal(t.T(), fmt.Errorf(invalidConfigError, -1, 10), err)
+}
+
+func (t *BlockPoolTest) TestInitBlockPoolForZeroMaxBlocks() {
+	_, err := InitBlockPool(10, 0)
+
+	require.NotNil(t.T(), err)
+	assert.Equal(t.T(), fmt.Errorf(invalidConfigError, 10, 0), err)
+}
+
+func (t *BlockPoolTest) TestInitBlockPoolForNegativeMaxBlocks() {
+	_, err := InitBlockPool(10, -1)
+
+	require.NotNil(t.T(), err)
+	assert.Equal(t.T(), fmt.Errorf(invalidConfigError, 10, -1), err)
+}
+
+// Represents when block is available on the blocksCh.
+func (t *BlockPoolTest) TestGetWhenBlockIsAvailableForReuse() {
+	bp, err := InitBlockPool(1024, 10)
+	require.Nil(t.T(), err)
+	// Creating a block with some data and send it to blockCh.
+	b := createBlock(1024)
+	content := []byte("hi")
+	err = b.Write(content)
+	require.Nil(t.T(), err)
+	// Validating the content of the block
+	output, err := io.ReadAll(b.Reader())
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), content, output)
+	bp.blocksCh <- b
+	// Setting totalBlocks same as maxBlocks to ensure no new blocks are created.
+	bp.totalBlocks = 10
+
+	block, err := bp.Get()
+
+	require.Nil(t.T(), err)
+	require.NotNil(t.T(), block)
+	// This ensures the block is reset.
+	assert.Equal(t.T(), int64(0), block.Size())
+}
+
+func (t *BlockPoolTest) TestGetWhenTotalBlocksIsLessThanThanMaxBlocks() {
+	bp, err := InitBlockPool(1024, 10)
+	require.Nil(t.T(), err)
+
+	block, err := bp.Get()
+
+	require.Nil(t.T(), err)
+	require.NotNil(t.T(), block)
+	assert.Equal(t.T(), int64(0), block.Size())
+}
+
+func (t *BlockPoolTest) TestCreateBlock() {
```

Comment:
This test is same as TestGetWhenTotalBlocksIsLessThanThanMaxBlocks
Shall we remove one?

Reply:
yep. Removed.